Blog post: Context Engineering for Agentic Systems (#515)
deepset-ai/haystack-home: content/blog/context-engineering/index.md

```python
from haystack import Pipeline
from haystack.components.embedders import SentenceTransformersTextEmbedder
from haystack.components.rankers import TransformersSimilarityRanker
from haystack.components.retrievers import InMemoryEmbeddingRetriever
from haystack.document_stores.in_memory import InMemoryDocumentStore

document_store = InMemoryDocumentStore()

rag = Pipeline()
rag.add_component("embedder", SentenceTransformersTextEmbedder())
# Retrieve 10 candidates, then rerank to the 3 most relevant
rag.add_component("retriever", InMemoryEmbeddingRetriever(document_store, top_k=10))
rag.add_component("ranker", TransformersSimilarityRanker(top_k=3))

rag.connect("embedder.embedding", "retriever.query_embedding")
rag.connect("retriever.documents", "ranker.documents")

result = rag.run({
    "embedder": {"text": "climate change"},
    "ranker": {"query": "climate change"},
})
# result["ranker"]["documents"] now contains at most 3 highly relevant chunks
```

> ### Summarisation and compaction

Author: Yeah, Anthropic uses it quite often: https://platform.claude.com/cookbook/tool-use-automatic-context-compaction

---
layout: blog-post
title: "Context Engineering for Agentic Systems: What Goes Into Your Agent's Mind"
description: A practical introduction to context engineering - what fills the LLM context window in agentic systems, why it matters, and how to keep it under control.
featured_image: thumbnail.png
featured_image_caption: The context window of an agentic system is not a blank slate - it is filled by many components, each competing for the model's attention.
images: ["blog/context-engineering/thumbnail.png", "blog/context-engineering/context-growth.png", "blog/context-engineering/context-breakdown.png"]
toc: True
date: 2026-03-17

> last_updated: 2026-03-17

Reviewer: We need to update these days later on

authors:
  - Kacper Lukawski

> canonical_url: https://www.deepset.ai/blog/context-engineering

Reviewer: I'm not sure if this is how we should use it 🤔

Author: I'll remove it, just copied the parameter from the "Build a Search Engine with GPT-3" post.

The practical consequence is that irrelevant or redundant content doesn't just waste space - it actively competes with the information that actually matters. A critical instruction buried under pages of tool outputs may receive less attention than if it had been sent alone. [Research from Anthropic](https://www.anthropic.com/engineering/effective-context-engineering-for-ai-agents) confirms this: models remain capable at longer contexts but show reduced precision for information retrieval and long-range reasoning compared to shorter ones. A million-token context window is not a free pass to include everything - it's a budget, and every token you add is a trade-off.

### The cost dimension

Most hosted LLMs charge per input token, which means every byte of context has a price tag. A single call with a 50,000-token context costs roughly 50× more than one with 1,000 tokens - and in an agentic loop that runs dozens of iterations, that multiplier compounds with every step. Context management is therefore not just a quality concern but a cost concern: a bloated context window can turn a cheap pipeline into an expensive one without producing any better answers.

> ---

Reviewer: ```suggestion
```

## What fills the context window in an agentic system

We've already touched on some of the components that fill an agent's context window - system prompts, tool definitions, retrieved documents. Let's map out the full picture, because the list is longer than many developers expect.

- **System prompt** - standing instructions, persona, constraints, output format. Usually fixed but can be large.
- **Conversation history** - the full back-and-forth between user and agent across the current session.

> - **Memory** - retrieved facts from past sessions or external knowledge stores. See also: [Good Listener: How Memory Enables Conversational Agents](/blog/memory-conversational-agents/).

Reviewer: I think linking to this notebook makes more sense here, the current blog article refers to conversation history:
```suggestion
- **Memory** - retrieved facts from past sessions or external knowledge stores. See also: [Using Mem0 Memory Store with Haystack Agents](https://haystack.deepset.ai/cookbook/memory_store_mem0).
```

# The agent is initially provided only with the search_tools tool
# and will use it to find relevant tools on demand.
result = agent.run(messages=[ChatMessage.from_user("What's the weather in Milan?")])
```

### Offloading notes (scratchpad / working memory)

An agent's intermediate reasoning - the chain of thoughts it builds up while working through a multi-step task - does not have to live inside the context window. A simple alternative is to give the agent two dedicated tools: one to write a note to an external store, and one to read notes back. Instead of accumulating its internal monologue in the prompt, the agent can offload conclusions, partial results, and reminders to storage and retrieve only what it needs at each step.

This keeps the context lean: rather than carrying the full trace of every intermediate thought, the agent holds a minimal working state and queries its own notes on demand. The pattern is especially useful for long-horizon tasks where the reasoning chain would otherwise grow without bound, and it has the side effect of making the agent's thinking inspectable and debuggable from outside the model.

> ---

Reviewer: ```suggestion
```
